mesonbuild/coredata.py

```diff
@@ -107,21 +110,15 @@ def set_value(self, newvalue: T.Any) -> None:
         self.value = self.validate_value(newvalue)
 
 class UserStringOption(UserOption[str]):
-    def __init__(self, description: str, value: T.Any, yielding: bool = DEFAULT_YIELDING,
-                 deprecated: T.Union[bool, str, T.Dict[str, str], T.List[str]] = False):
-        super().__init__(description, None, yielding, deprecated)
-        self.set_value(value)
 
     def validate_value(self, value: T.Any) -> str:
         if not isinstance(value, str):
             raise MesonException('Value "%s" for string option is not a string.' % str(value))
         return value
 
 class UserBooleanOption(UserOption[bool]):
-    def __init__(self, description: str, value, yielding: bool = DEFAULT_YIELDING,
-                 deprecated: T.Union[bool, str, T.Dict[str, str], T.List[str]] = False):
-        super().__init__(description, [True, False], yielding, deprecated)
-        self.set_value(value)
+
+    choices: T.List[bool] = field(default_factory=lambda: [True, False])
```

Comment:
Do we need a factory field here (and elsewhere)? AFAICT we never modify choices in place, only assign them as `= newthing`, and never for boolean or umask anyway. Can `typing_extensions.Final` help here?

Reply:
Particularly given this is a static choice on the option type, I feel like this should be a ClassVar, but apparently you cannot mix ClassVar and Final: https://bugs.python.org/issue45384

Pity. But personally I'd still go with a ClassVar I guess.

Reply:
This is all really annoying to solve because with dataclasses you can't forward declare an attribute. I've done my best here by adding an invisible attribute for `choices`, and then overriding it all of the subclasses. The problem is that without it being defined in UserOption, UserOption can't be used as a generic interface, and mypy says "hey, that dict of [str: UserOption] doesn't have .choices!"

Reply:
Actually, I realized I'm wrong, you *can* forward declare them, you just have to do it the old-school way. 
```python
@dataclass
class C:
    x: int
    if typing.TYPE_CHECKING:
        y: str
```

But it doesn't help us in this case